frontend/src/metabase/query_builder/containers/QueryBuilder.tsx

```diff
@@ -213,7 +251,7 @@ function QueryBuilder(props) {
     () => _.debounce(forceUpdate, 400),
     [forceUpdate],
   );
-  const timeout = useRef(null);
+  const timeout = useRef<NodeJS.Timeout>();
```

Comment:
Looks like nodejs typings are interfering with dom typings.
But it's `number` in the browser: https://github.com/microsoft/TypeScript/blob/2c23bea/src/lib/dom.generated.d.ts#L28698

```suggestion
  const timeout = useRef<number>();
```

Reply:
I followed the suggestion but it seems like at 
```
timeout.current = setTimeout(forceUpdateDebounced, 300);
```

we get the whole error of `number` not assignable to `NodeJS.Timeout`. I looked at this [StackOverflow thread](https://stackoverflow.com/a/60247753) and we have two options:

- We _could_ use `window.setTimeout` which will use the TS type that you've linked to. 
- use `useRef<ReturnType<typeof setTimeout>>()` but that still seems to use the `NodeJS.Timeout` under the hood

Reply:
It seems that we're using almost all possible approaches in the codebase:
- `any` https://github.com/metabase/metabase/blob/92d615073a3ba31710e26be9f2e83141352d1d1c/frontend/src/metabase/nav/containers/MainNavbar/SidebarItems/SidebarCollectionLink.tsx#L48
- `NodeJS.Timeout` https://github.com/metabase/metabase/blob/d897f2016fa86e88e1e4936a53ef14a64c803db6/frontend/src/metabase/hooks/use-action-button-label/use-action-button-label.ts#L18
- `ReturnType<typeof setTimeout>` https://github.com/metabase/metabase/blob/92d615073a3ba31710e26be9f2e83141352d1d1c/frontend/src/metabase/ui/components/utils/DelayGroup/DelayGroup.tsx#L35

Now we just need `window.setTimeout` to get the completionist achievement.

---

The only one that gives proper typing is `window.setTimeout`.